python/langsmith/client.py

```diff
@@ -4664,6 +4664,538 @@ def _evaluate_strings(
+    ) -> Any:
+        """Pull a prompt and return it as a LangChain PromptTemplate.
+
+        This method requires `langchain_core`.
+
+        Args:
+            prompt_identifier (str): The identifier of the prompt.
+
+        Returns:
+            Any: The prompt object in the specified format.
+        """
+        try:
+            from langchain_core.load.load import loads
+            from langchain_core.prompts import BasePromptTemplate
+        except ImportError:
+            raise ImportError(
+                "The client.pull_prompt function requires the langchain_core"
+                "package to run.\nInstall with `pip install langchain_core`"
+            )
+
+        prompt_object = self.pull_prompt_object(
+            prompt_identifier, include_model=include_model
+        )
+        prompt = loads(json.dumps(prompt_object.manifest))
+
+        if isinstance(prompt, BasePromptTemplate) or isinstance(
+            prompt.first, BasePromptTemplate
+        ):
+            prompt_template = (
+                prompt if isinstance(prompt, BasePromptTemplate) else prompt.first
+            )
+            if prompt_template.metadata is None:
+                prompt_template.metadata = {}
+            prompt_template.metadata.update(
+                {
+                    "lc_hub_owner": prompt_object.owner,
+                    "lc_hub_repo": prompt_object.repo,
+                    "lc_hub_commit_hash": prompt_object.commit_hash,
+                }
+            )
+
+        return prompt
+
+    def push_prompt(
+        self,
+        prompt_identifier: str,
+        *,
+        object: Optional[Any] = None,
+        parent_commit_hash: Optional[str] = "latest",
+        is_public: bool = False,
+        description: Optional[str] = "",
+        readme: Optional[str] = "",
+        tags: Optional[List[str]] = [],
+    ) -> str:
+        """Push a prompt to the LangSmith API.
+
+        Can be used to update prompt metadata or prompt content.
+
+        If the prompt does not exist, it will be created.
+        If the prompt exists, it will be updated.
+
+        Args:
+            prompt_identifier (str): The identifier of the prompt.
+            object (Optional[Any]): The LangChain object to push.
+            parent_commit_hash (Optional[str]): The parent commit hash.
+              Defaults to "latest".
+            is_public (bool): Whether the prompt should be public. Defaults to False.
+            description (Optional[str]): A description of the prompt.
+              Defaults to an empty string.
+            readme (Optional[str]): A readme for the prompt.
+              Defaults to an empty string.
+            tags (Optional[List[str]]): A list of tags for the prompt.
+              Defaults to an empty list.
+
+        Returns:
+            str: The URL of the prompt.
+
+        """
+        # Create or update prompt metadata
+        if self._prompt_exists(prompt_identifier):
```

Comment:
would prefer a single upsert endpoint eventually

Reply:
yeah that makes sense